File: ios/Sources/AudioPlayerPlugin/AudioMetadata.swift
import Capacitor
import UserNotifications

// AudioMetadata fetches a 2-minute lookahead of upcoming track changes from
// soundz-good's /metadata-upcoming endpoint every 60 s, holds the items in
// a local queue, and applies each to the OS Now Playing display at its
// own from_us via a single self-rearming apply timer. AudioSource still
// reads the current track via the same property surface (artist, title,
// trackId, …) — those are now computed from the queue's currently-active
// entry rather than the single track payload of the old endpoint.
//
// Queue is dropped on stop/play/skip (the old timeline is stale); the
// next refresh repopulates it. Apply-timer fires also drop entries whose
// to_us is in the past, so the queue stays small.
public class AudioMetadata {
    // Silent shade-only ad-indicator notification, mirror of the Android
    // implementation. Gated on `is_ad` AND a non-empty `target_url`: the
    // backend now populates target_url for any track with a click destination
    // (ads and songs alike), so is_ad is what distinguishes "show the ad
    // shade" from a song's in-app click-through.
    private static let adNotificationId = "now_playing_passive"

    private static let refreshIntervalSeconds: Int = 60

    var channelId: String = ""
    var updateUrl: String

    private var queue: [UpcomingMetadataItem] = []
    private let queueLock = NSLock()

    private var refreshTimer: DispatchSourceTimer?
    private var applyTimer: DispatchSourceTimer?

    private var updateCallback: (() -> Void)!
    private var pluginOwner: AudioPlayerPlugin?

    private var pollerActive: Bool = false
    private var previousNotifiedTrackId: String = ""

    // Current-item accessors: AudioSource and the host plugin read these
    // identically to before; the underlying storage is now the queue.
    var trackId: String { currentItem()?.id ?? "" }
    var artist: String { currentItem()?.artist ?? "" }
    var title: String { currentItem()?.title ?? "" }
    var album: String { currentItem()?.album ?? "" }
    var imageUrl: String { currentItem()?.imageUrl ?? "" }
    var maySkip: Bool { currentItem()?.maySkip ?? true }
    var isAd: Bool { currentItem()?.isAd ?? false }
    var targetUrl: String { currentItem()?.targetUrl ?? "" }

    public init(updateUrl: String) {
        self.updateUrl = updateUrl
    }

    public func setPluginOwner(pluginOwner: AudioPlayerPlugin) -> Self {
        self.pluginOwner = pluginOwner

        return self
    }

    public func setUpdateCallback(callback: @escaping () -> Void) -> Self {
        self.updateCallback = callback

        return self
    }

    public func startUpdater() {
        if !hasUpdateUrl() || refreshTimer != nil {
            return
        }

        pollerActive = true
        let timer = DispatchSource.makeTimerSource(
            queue: DispatchQueue.global(qos: .background)
        )
        // 1s initial delay lets the stream warm up server-side (the
        // forwarder's first items reach Redis) before the first poll, so
        // the lockscreen picks up real metadata on the very first request.
        timer.schedule(
            deadline: .now() + 1.0,
            repeating: .seconds(Self.refreshIntervalSeconds),
            leeway: .milliseconds(250)
        )
        timer.setEventHandler { [weak self] in
            self?.makeUpdateRequest()
        }
        refreshTimer = timer
        timer.activate()
    }

    public func stopUpdater() {
        if refreshTimer == nil {
            return
        }

        pollerActive = false
        clearAdNotification()
        refreshTimer?.cancel()
        refreshTimer = nil
        applyTimer?.cancel()
        applyTimer = nil

        // Drop the queue: the next session's timeline is unrelated.
        queueLock.lock()
        queue = []
        queueLock.unlock()

        // One delayed final poll so the OS audio controls catch up to the
        // server's post-disconnect state (e.g. soundz-good returning channel
        // metadata after the stream connection closes and Deregister fires).
        if hasUpdateUrl() {
            DispatchQueue.global(qos: .background).asyncAfter(deadline: .now() + 1.0) {
                self.makeUpdateRequest()
            }
        }
    }

    public func hasUpdateUrl() -> Bool {
        return !updateUrl.isEmpty
    }

    // Force a one-shot refresh — same semantics as before, used by the
    // foreground-resume path on the JS side.
    public func updateMetadataByUrl() {
        makeUpdateRequest()
    }

    private func currentItem() -> UpcomingMetadataItem? {
        let nowUs = AudioMetadata.nowUs()
        queueLock.lock()
        defer { queueLock.unlock() }
        for item in queue {
            if item.fromUs <= nowUs && nowUs <= item.toUs {
                return item
            }
        }
        return nil
    }

    private static func nowUs() -> Int64 {
        return Int64(Date().timeIntervalSince1970 * 1_000_000)
    }

    private func makeUpdateRequest() {
        guard let updateUrl = URL(string: self.updateUrl) else {
            print("Update metadata URL is invalid")
            return
        }

        var updateRequest = URLRequest(url: updateUrl)
        updateRequest.addValue("application/json", forHTTPHeaderField: "Accept")

        URLSession.shared.dataTask(with: updateRequest) { (data, response, error) in
            if let error = error {
                print(error)
                return
            }

            guard let httpResponse = response as? HTTPURLResponse else {
                print("The metadata update server response is invalid")
                return
            }

            guard (200...299).contains(httpResponse.statusCode) else {
                print(
                    "The metadata update server returned a non-2xx status code: \(httpResponse.statusCode)"
                )
                return
            }

            do {
                guard
                    let json = (try JSONSerialization.jsonObject(with: data!)) as? [String: Any]
                else {
                    print("The metadata update data could not be parsed as JSON")
                    return
                }

                self.channelId = json["channel_id"] as? String ?? ""
                var newQueue: [UpcomingMetadataItem] = []
                if let items = json["items"] as? [[String: Any]] {
                    for item in items {
                        newQueue.append(UpcomingMetadataItem(json: item))
                    }
                }
                newQueue.sort { $0.fromUs < $1.fromUs }

                self.queueLock.lock()
                self.queue = newQueue
                self.queueLock.unlock()

                // Apply the now-active entry immediately (the refresh
                // landed mid-group; the previous queue's current item may
                // have differed) and arm the next-item timer.
                DispatchQueue.main.async {
                    self.fireMetadataUpdated()
                }
                self.scheduleNextApply()
            } catch {
                print(
                    "An error occurred trying to get updated metadata: \(error.localizedDescription)"
                )
            }
        }.resume()
    }

    // scheduleNextApply prunes outdated entries and arms a one-shot timer
    // for the next item whose from_us is in the future. The timer's fire
    // handler invokes the OS-update callback and self-rearms for the item
    // after that — so there's only ever one pending DispatchSourceTimer.
    private func scheduleNextApply() {
        applyTimer?.cancel()
        applyTimer = nil

        let nowUs = AudioMetadata.nowUs()
        queueLock.lock()
        queue = queue.filter { $0.toUs >= nowUs }
        let next = queue.first(where: { $0.fromUs > nowUs })
        queueLock.unlock()

        guard let next = next else { return }

        let delaySeconds = max(0.0, Double(next.fromUs - nowUs) / 1_000_000.0)
        let timer = DispatchSource.makeTimerSource(
            queue: DispatchQueue.global(qos: .background)
        )
        timer.schedule(deadline: .now() + delaySeconds, leeway: .milliseconds(10))
        timer.setEventHandler { [weak self] in
            guard let self = self else { return }
            DispatchQueue.main.async {
                self.fireMetadataUpdated()
            }
            self.scheduleNextApply()
        }
        applyTimer = timer
        timer.activate()
    }

    private func fireMetadataUpdated() {
        syncAdNotification()
        if updateCallback != nil {
            updateCallback()
        }
    }

    private func syncAdNotification() {
        guard pollerActive else { return }

        let center = UNUserNotificationCenter.current()

        // Same gate as Android: only show the shade notification for ads
        // with a click destination. Songs may also carry target_url but get
        // their click-through from the in-app UI, not a shade notification.
        guard isAd && !targetUrl.isEmpty else {
            center.removeDeliveredNotifications(withIdentifiers: [Self.adNotificationId])
            center.removePendingNotificationRequests(withIdentifiers: [Self.adNotificationId])
            previousNotifiedTrackId = ""
            return
        }

        // Skip repeat polls of the same track to avoid flicker. Each
        // re-post would trigger a brief shade refresh even at .passive.
        guard trackId != previousNotifiedTrackId else { return }
        previousNotifiedTrackId = trackId

        // Capture state for the closure — the URLSession callback runs on a
        // background queue and the metadata fields could change before we
        // post.
        let snapshotTitle = title.isEmpty ? "Now playing" : title
        let snapshotArtist = artist
        let snapshotTargetUrl = targetUrl
        let snapshotImageUrl = imageUrl

        fetchAttachmentURL(snapshotImageUrl) { attachmentURL in
            let content = UNMutableNotificationContent()
            content.title = snapshotTitle
            content.body = snapshotArtist
            content.sound = nil
            content.interruptionLevel = .passive  // shade-only, no banner
            content.userInfo = ["target_url": snapshotTargetUrl]

            if let url = attachmentURL,
               let attachment = try? UNNotificationAttachment(
                   identifier: "art",
                   url: url,
                   options: nil
               ) {
                content.attachments = [attachment]
            }

            let request = UNNotificationRequest(
                identifier: Self.adNotificationId,
                content: content,
                trigger: nil
            )
            center.add(request) { error in
                if let error = error {
                    print("Failed to post ad-indicator notification: \(error)")
                }
            }
        }
    }

    private func clearAdNotification() {
        let center = UNUserNotificationCenter.current()
        center.removeDeliveredNotifications(withIdentifiers: [Self.adNotificationId])
        center.removePendingNotificationRequests(withIdentifiers: [Self.adNotificationId])
        previousNotifiedTrackId = ""
    }

    // UNNotificationAttachment requires a local file URL. Mirror
    // AudioSource.downloadNowPlayingIcon's URLSession pattern but write to a
    // unique tmp file so back-to-back ads don't collide. Returns nil for
    // empty/invalid URLs, network errors, or write failures — caller posts
    // the notification without an attachment. Image is optional, never fatal.
    private func fetchAttachmentURL(
        _ urlString: String,
        completion: @escaping (URL?) -> Void
    ) {
        guard !urlString.isEmpty,
              let url = URL(string: urlString),
              let scheme = url.scheme,
              ["http", "https"].contains(scheme.lowercased())
        else {
            completion(nil)
            return
        }

        URLSession.shared.dataTask(with: url) { data, _, _ in
            guard let data = data else {
                completion(nil)
                return
            }
            let tmp = FileManager.default.temporaryDirectory
                .appendingPathComponent(UUID().uuidString + ".jpg")
            do {
                try data.write(to: tmp)
                completion(tmp)
            } catch {
                completion(nil)
            }
        }.resume()
    }
}

struct UpcomingMetadataItem {
    let fromUs: Int64
    let toUs: Int64
    let id: String
    let artist: String
    let title: String
    let album: String
    let imageUrl: String
    let maySkip: Bool
    let isAd: Bool
    let targetUrl: String

    init(json: [String: Any]) {
        self.fromUs = (json["from_us"] as? NSNumber)?.int64Value ?? 0
        self.toUs = (json["to_us"] as? NSNumber)?.int64Value ?? 0
        self.id = json["id"] as? String ?? ""
        self.artist = json["artist"] as? String ?? ""
        self.title = json["title"] as? String ?? ""
        self.album = json["album"] as? String ?? ""
        self.imageUrl = json["image_url"] as? String ?? ""
        self.maySkip = json["may_skip"] as? Bool ?? true
        self.isAd = json["is_ad"] as? Bool ?? false
        self.targetUrl = json["target_url"] as? String ?? ""
    }
}

File: android/src/main/java/us/mediagrid/capacitorjs/plugins/nativeaudio/AudioMetadata.java
package us.mediagrid.capacitorjs.plugins.nativeaudio;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import androidx.core.app.NotificationCompat;
import com.getcapacitor.JSObject;
import com.getcapacitor.plugin.util.HttpRequestHandler;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONObject;

// AudioMetadata fetches a 2-minute lookahead of upcoming track changes from
// soundz-good's /metadata-upcoming endpoint every 60 s, holds the items in
// a local queue, and applies each to the OS MediaSession at its own from_us
// via a self-rearming apply Handler. AudioSource and AudioPlayerService still
// read the current track via the same field surface (artist, title, trackId,
// …) — those fields are written by the apply timer whenever a new group
// becomes current.
//
// Queue is dropped on stop/play/skip (the old timeline is stale); the next
// refresh repopulates it. Apply-timer fires also drop entries whose to_us
// is in the past, so the queue stays small.
public class AudioMetadata {

    private static final String TAG = "AudioMetadata";

    // Silent shade-only ad-indicator notification. Gated on `is_ad` AND a
    // non-empty `target_url`: the backend now populates target_url for any
    // track with a click destination (ads and songs alike), so is_ad is what
    // distinguishes "show the ad shade" from a song's in-app click-through.
    private static final String AD_CHANNEL_ID = "now_playing_passive";
    private static final int AD_NOTIFICATION_ID = 9912;

    private static final int REFRESH_INTERVAL_MS = 60 * 1000;

    public String channelId = "";
    public String trackId = "";
    public String artist = "";
    public String title = "";
    public String album = "";
    // Default to the bundled favicon as the lockscreen artwork until the
    // first /metadata-upcoming poll lands a real image_url.
    public String imageUrl = "favicon.png";
    public boolean maySkip = true;
    public boolean isAd = false;
    public String targetUrl = "";

    public String updateUrl;

    private final List<UpcomingMetadataItem> queue = new ArrayList<>();

    private Handler refreshHandler = null;
    private Runnable refreshRunner = null;
    private Handler applyHandler = null;
    private Runnable applyRunner = null;
    private Runnable updateCallback = null;

    private AudioPlayerPlugin pluginOwner;

    private boolean pollerActive = false;
    private String previousNotifiedTrackId = "";

    AudioMetadata(String updateUrl) {
        this.updateUrl = updateUrl;
    }

    public AudioMetadata setPluginOwner(AudioPlayerPlugin plugin) {
        pluginOwner = plugin;

        return this;
    }

    public AudioMetadata setUpdateCallBack(Runnable callback) {
        updateCallback = callback;

        return this;
    }

    public void startUpdater() {
        if (!hasUpdateUrl() || refreshHandler != null) {
            return;
        }

        pollerActive = true;
        refreshHandler = new Handler(Looper.getMainLooper());
        applyHandler = new Handler(Looper.getMainLooper());
        refreshRunner = new Runnable() {
            @Override
            public void run() {
                makeUpdateRequest(() -> {
                    if (refreshHandler != null) {
                        refreshHandler.postDelayed(this, REFRESH_INTERVAL_MS);
                    }
                });
            }
        };

        // 1s delay lets the stream warm up server-side (the forwarder's
        // first items reach Redis) before the first poll, so the lockscreen
        // picks up real metadata on the very first request.
        refreshHandler.postDelayed(refreshRunner, 1000);
    }

    public void stopUpdater() {
        if (refreshHandler == null) {
            return;
        }

        pollerActive = false;
        clearTrackNotification();
        refreshHandler.removeCallbacks(refreshRunner);
        refreshHandler = null;
        refreshRunner = null;
        if (applyHandler != null && applyRunner != null) {
            applyHandler.removeCallbacks(applyRunner);
        }
        applyHandler = null;
        applyRunner = null;

        // Drop the queue: the next session's timeline is unrelated.
        synchronized (queue) {
            queue.clear();
        }

        // One delayed final poll so the OS audio controls catch up to the
        // server's post-disconnect state (e.g. soundz-good returning channel
        // metadata after the stream connection closes and Deregister fires).
        if (hasUpdateUrl()) {
            new Handler(Looper.getMainLooper()).postDelayed(
                () -> makeUpdateRequest(null),
                1000
            );
        }
    }

    public boolean hasUpdateUrl() {
        return updateUrl != null && !updateUrl.isEmpty();
    }

    // Force a one-shot refresh — same semantics as before, used by the
    // foreground-resume path on the JS side.
    public void updateMetadataByUrl(Runnable requeueCallback) {
        makeUpdateRequest(requeueCallback);
    }

    private void makeUpdateRequest(Runnable requeueCallback) {
        if (pluginOwner.executorService.isShutdown()) {
            return;
        }

        pluginOwner.executorService.submit(() -> {
            Log.i(TAG, "poll firing for URL=" + updateUrl);
            try {
                fetchAndApplyWindow();
            } catch (Exception ex) {
                Log.e(TAG, "There was an error running the metadata update", ex);
            } finally {
                // Always requeue — a single failed poll (e.g. 404 during
                // stream startup race) must not kill the poller permanently.
                if (requeueCallback != null) {
                    requeueCallback.run();
                }
            }
        });
    }

    private void fetchAndApplyWindow() {
        Log.i(TAG, "Getting metadata from URL " + updateUrl);
        HttpURLConnection urlConnection = null;

        try {
            URL url = new URL(updateUrl);
            urlConnection = (HttpURLConnection) url.openConnection();
            urlConnection.setRequestProperty("Accept", "application/json");

            InputStream errorStream = urlConnection.getErrorStream();
            if (errorStream != null) {
                Log.e(
                    TAG,
                    String.format(
                        "The metadata update server returned a status of %s with the message %s",
                        urlConnection.getResponseCode(),
                        HttpRequestHandler.readStreamAsString(errorStream)
                    )
                );
                return;
            }

            JSObject json = new JSObject(
                HttpRequestHandler.readStreamAsString(urlConnection.getInputStream())
            );

            String newChannelId = stringOrEmpty(json.getString("channel_id"));
            List<UpcomingMetadataItem> newQueue = new ArrayList<>();
            JSONArray items = json.optJSONArray("items");
            if (items != null) {
                for (int i = 0; i < items.length(); i++) {
                    JSONObject obj = items.optJSONObject(i);
                    if (obj == null) continue;
                    newQueue.add(UpcomingMetadataItem.fromJson(obj));
                }
            }
            Collections.sort(newQueue, (a, b) -> Long.compare(a.fromUs, b.fromUs));

            synchronized (queue) {
                queue.clear();
                queue.addAll(newQueue);
            }
            channelId = newChannelId;

            // Apply the now-active entry immediately (the refresh landed
            // mid-group; the previous queue's current item may have
            // differed) and arm the apply timer for the next transition.
            new Handler(Looper.getMainLooper()).post(() -> {
                fireMetadataUpdated();
                scheduleNextApply();
            });
        } catch (Exception ex) {
            Log.e(TAG, "An error occurred trying to get updated metadata", ex);
        } finally {
            if (urlConnection != null) {
                urlConnection.disconnect();
            }
        }
    }

    // scheduleNextApply prunes outdated entries and arms a one-shot post
    // for the next item whose from_us is in the future. The post handler
    // fires fireMetadataUpdated() and re-arms — so there's only ever one
    // pending applyHandler callback.
    private void scheduleNextApply() {
        if (applyHandler == null) return;
        if (applyRunner != null) {
            applyHandler.removeCallbacks(applyRunner);
            applyRunner = null;
        }

        long nowUs = nowUs();
        UpcomingMetadataItem next;
        synchronized (queue) {
            Iterator<UpcomingMetadataItem> it = queue.iterator();
            while (it.hasNext()) {
                if (it.next().toUs < nowUs) {
                    it.remove();
                }
            }
            next = null;
            for (UpcomingMetadataItem item : queue) {
                if (item.fromUs > nowUs) {
                    next = item;
                    break;
                }
            }
        }
        if (next == null) return;

        long delayMs = Math.max(0, (next.fromUs - nowUs) / 1000);
        applyRunner = () -> {
            fireMetadataUpdated();
            scheduleNextApply();
        };
        applyHandler.postDelayed(applyRunner, delayMs);
    }

    private void fireMetadataUpdated() {
        applyCurrentItem();
        syncTrackNotification();
        if (updateCallback != null) {
            updateCallback.run();
        }
    }

    // applyCurrentItem copies the queue entry covering "now" into the public
    // fields AudioSource / AudioPlayerService read. When no entry covers now,
    // leave the last-applied values in place — the lockscreen continues
    // showing the previous track until the next refresh closes the gap.
    private void applyCurrentItem() {
        long nowUs = nowUs();
        UpcomingMetadataItem current = null;
        synchronized (queue) {
            for (UpcomingMetadataItem item : queue) {
                if (item.fromUs <= nowUs && nowUs <= item.toUs) {
                    current = item;
                    break;
                }
            }
        }
        if (current == null) {
            return;
        }
        trackId = current.id;
        artist = current.artist;
        title = current.title;
        album = current.album;
        imageUrl = current.imageUrl;
        maySkip = current.maySkip;
        isAd = current.isAd;
        targetUrl = current.targetUrl;
    }

    private static long nowUs() {
        return System.currentTimeMillis() * 1000L;
    }

    private static String stringOrEmpty(String s) {
        return s == null ? "" : s;
    }

    private Context notificationContext() {
        if (pluginOwner == null) return null;
        Context ctx = pluginOwner.getContext();
        return ctx == null ? null : ctx.getApplicationContext();
    }

    private void ensureNotificationChannel(Context ctx, NotificationManager nm) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return;
        if (nm.getNotificationChannel(AD_CHANNEL_ID) != null) return;
        NotificationChannel channel = new NotificationChannel(
            AD_CHANNEL_ID,
            "Now playing",
            NotificationManager.IMPORTANCE_LOW
        );
        channel.setDescription("Quietly indicates the currently playing track. Silent, no banner.");
        channel.setSound(null, null);
        channel.enableVibration(false);
        nm.createNotificationChannel(channel);
    }

    private void syncTrackNotification() {
        if (!pollerActive) return;

        Context ctx = notificationContext();
        if (ctx == null) return;
        NotificationManager nm = (NotificationManager) ctx.getSystemService(Context.NOTIFICATION_SERVICE);
        if (nm == null) return;

        // Gate is is_ad AND a click destination — songs may also carry
        // target_url but get their click-through from the in-app UI, not a
        // shade notification.
        if (!isAd || targetUrl.isEmpty()) {
            nm.cancel(AD_NOTIFICATION_ID);
            previousNotifiedTrackId = "";
            return;
        }

        if (trackId.equals(previousNotifiedTrackId)) return;

        ensureNotificationChannel(ctx, nm);
        nm.cancel(AD_NOTIFICATION_ID);
        previousNotifiedTrackId = trackId;

        // Tap → ClickLink endpoint via system browser. The backend records
        // the click against (track, asset, campaign) then 302-redirects to
        // the asset's landing page.
        Intent view = new Intent(Intent.ACTION_VIEW, Uri.parse(targetUrl));
        view.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        PendingIntent contentPi = PendingIntent.getActivity(
            ctx,
            2,
            view,
            PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT
        );

        Bitmap art = fetchBitmap(imageUrl);

        NotificationCompat.Builder builder = new NotificationCompat.Builder(ctx, AD_CHANNEL_ID)
            .setSmallIcon(R.drawable.media3_notification_small_icon)
            .setContentTitle(title.isEmpty() ? "Now playing" : title)
            .setContentText(artist)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .setSilent(true)
            .setOnlyAlertOnce(true)
            .setContentIntent(contentPi)
            .setAutoCancel(true);

        if (art != null) {
            builder.setLargeIcon(art)
                .setStyle(new NotificationCompat.BigPictureStyle()
                    .bigPicture(art)
                    .bigLargeIcon((Bitmap) null));
        }

        nm.notify(AD_NOTIFICATION_ID, builder.build());
        Log.i(TAG, "syncTrackNotification: posted for trackId=" + trackId
            + " withArt=" + (art != null));
    }

    private Bitmap fetchBitmap(String url) {
        if (url == null || url.isEmpty()) return null;
        if (!url.startsWith("http://") && !url.startsWith("https://")) return null;
        HttpURLConnection conn = null;
        try {
            conn = (HttpURLConnection) new URL(url).openConnection();
            conn.setRequestProperty("Accept", "image/*");
            try (InputStream in = conn.getInputStream()) {
                return BitmapFactory.decodeStream(in);
            }
        } catch (Exception ex) {
            Log.w(TAG, "fetchBitmap failed for " + url + ": " + ex.getMessage());
            return null;
        } finally {
            if (conn != null) conn.disconnect();
        }
    }

    private void clearTrackNotification() {
        Context ctx = notificationContext();
        if (ctx == null) return;
        NotificationManager nm = (NotificationManager) ctx.getSystemService(Context.NOTIFICATION_SERVICE);
        if (nm != null) nm.cancel(AD_NOTIFICATION_ID);
        previousNotifiedTrackId = "";
    }

    private static class UpcomingMetadataItem {
        final long fromUs;
        final long toUs;
        final String id;
        final String artist;
        final String title;
        final String album;
        final String imageUrl;
        final boolean maySkip;
        final boolean isAd;
        final String targetUrl;

        UpcomingMetadataItem(long fromUs, long toUs, String id, String artist, String title,
                             String album, String imageUrl, boolean maySkip, boolean isAd,
                             String targetUrl) {
            this.fromUs = fromUs;
            this.toUs = toUs;
            this.id = id;
            this.artist = artist;
            this.title = title;
            this.album = album;
            this.imageUrl = imageUrl;
            this.maySkip = maySkip;
            this.isAd = isAd;
            this.targetUrl = targetUrl;
        }

        static UpcomingMetadataItem fromJson(JSONObject obj) {
            return new UpcomingMetadataItem(
                obj.optLong("from_us", 0),
                obj.optLong("to_us", 0),
                obj.optString("id", ""),
                obj.optString("artist", ""),
                obj.optString("title", ""),
                obj.optString("album", ""),
                obj.optString("image_url", ""),
                obj.optBoolean("may_skip", true),
                obj.optBoolean("is_ad", false),
                obj.optString("target_url", "")
            );
        }
    }
}
